d2/src/main/java/com/linkedin/d2/balancer/properties/ClusterPropertiesJsonSerializer.java

```diff
@@ -110,7 +116,14 @@ public ClusterProperties fromMap(Map<String, Object> map) {
         mapGetOrDefault(distributionStrategyMap, PropertyKeys.TARGET_HOSTS_STRATEGY_PROPERTIES, Collections.emptyMap()),
         mapGetOrDefault(distributionStrategyMap, PropertyKeys.TARGET_APPLICATIONS_STRATEGY_PROPERTIES, Collections.emptyMap()));
     }
-    return new ClusterStoreProperties(stableConfigs, canaryConfigs, distributionStrategy);
+    if (clusterFailoutMap != null && !clusterFailoutMap.isEmpty())
+    {
+      clusterFailoutProperties = new ClusterFailoutProperties(
```

Comment:
Will this failout property be used on d2-c-s as well? (I guess it will be, for writing the API and impl to handle a failout?) If so, it should be defined with a .pdl file so d2-c-s will share the generated D2 object.

And we usually have a converter to convert between property map to D2 object. Please reference D2CanaryDistributionStrategy.pdl and [CanaryDistributionStrategyConverter.java](https://github.com/linkedin/rest.li/pull/755/files#diff-0fd2cbc00acbaa5c3a8e8f06194e797c542a99fcfb1c8646c9a62eecc214aefd).

Reply:
pdl will live in another library, as per our conversation.